types/any/minimal_interface_registry.go

```diff
@@ -0,0 +1,222 @@
+package types
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/cosmos/gogoproto/proto"
+)
+
+var protoMessageType = reflect.TypeOf((*proto.Message)(nil)).Elem()
+
+// AnyUnpacker is an interface which allows safely unpacking types packed
+// in Any's against a whitelist of registered types
+type AnyUnpacker interface {
+	// UnpackAny unpacks the value in any to the interface pointer passed in as
+	// iface. Note that the type in any must have been registered in the
+	// underlying whitelist registry as a concrete type for that interface
+	// Ex:
+	//    var msg sdk.Msg
+	//    err := cdc.UnpackAny(any, &msg)
+	//    ...
+	UnpackAny(any *Any, iface interface{}) error
+}
+
+// InterfaceRegistry provides a mechanism for registering interfaces and
+// implementations that can be safely unpacked from Any
+type InterfaceRegistry interface {
+	AnyUnpacker
+
+	// RegisterInterface associates protoName as the public name for the
+	// interface passed in as iface. This is to be used primarily to create
+	// a public facing registry of interface implementations for clients.
+	// protoName should be a well-chosen public facing name that remains stable.
+	// RegisterInterface takes an optional list of impls to be registered
+	// as implementations of iface.
+	//
+	// Ex:
+	//   registry.RegisterInterface("cosmos.base.v1beta1.Msg", (*sdk.Msg)(nil))
+	RegisterInterface(protoName string, iface interface{}, impls ...proto.Message)
+
+	// RegisterImplementations registers impls as concrete implementations of
+	// the interface iface.
+	//
+	// Ex:
+	//  registry.RegisterImplementations((*sdk.Msg)(nil), &MsgSend{}, &MsgMultiSend{})
+	RegisterImplementations(iface interface{}, impls ...proto.Message)
+}
+
+// UnpackInterfacesMessage is meant to extend protobuf types (which implement
+// proto.Message) to support a post-deserialization phase which unpacks
+// types packed within Any's using the whitelist provided by AnyUnpacker
+type UnpackInterfacesMessage interface {
+	// UnpackInterfaces is implemented in order to unpack values packed within
+	// Any's using the AnyUnpacker. It should generally be implemented as
+	// follows:
+	//   func (s *MyStruct) UnpackInterfaces(unpacker AnyUnpacker) error {
+	//		var x AnyInterface
+	//		// where X is an Any field on MyStruct
+	//		err := unpacker.UnpackAny(s.X, &x)
+	//		if err != nil {
+	//			return nil
+	//		}
+	//		// where Y is a field on MyStruct that implements UnpackInterfacesMessage itself
+	//		err = s.Y.UnpackInterfaces(unpacker)
+	//		if err != nil {
+	//			return nil
+	//		}
+	//		return nil
+	//	 }
+	UnpackInterfaces(unpacker AnyUnpacker) error
+}
+
+type interfaceRegistry struct {
+	interfaceNames map[string]reflect.Type
+	interfaceImpls map[reflect.Type]interfaceMap
+	implInterfaces map[reflect.Type]reflect.Type
+	typeURLMap     map[string]reflect.Type
+}
+
+type interfaceMap = map[string]reflect.Type
+
+// NewInterfaceRegistry returns a new InterfaceRegistry
+func NewInterfaceRegistry() InterfaceRegistry {
```

Comment:
can we move that in `_test` or in an internal package?

Reply:
I did that and I ended up in a cycle import hell ... I'll see if can do it in a different way